### ISSUE TITLE
[Debugging] Looking for tips on how to trace why touch events aren't delivered to a control node.

### ISSUE THREAD

**Issue by rainypixels** [facebook/AsyncDisplayKit#2448]
This is a long shot "issue", but I'm keeping my fingers crossed that someone here has some tribal knowledge that might save me. 

I integrated ASDK beta 2 (`master`) into my app and began the process of migrating my clunky self-sizing `UITableView` instances to `ASCollectionNode` instances a few days ago. Everything was going really well until a couple of days ago when I added an `ASButtonNode` to one of my `ASCellNode` subclasses. It laid out perfectly, but when I tapped it... nothing. After working through the basic debugging paths — ensuring my layoutspec was sizing the node correctly, stripping my cells down to only the button, and a hundred other things — I caved and created a sample to isolate the problem. 

Fast-forward to today, I have a sample that works perfectly: the button taps work. The structure of the sample mimics that of my original app: roughly, a `UIPageViewController` that loads two `UIViewControllers` with a single view each that houses the `UICollectionNode` view as a subview. I even use demonic storyboards that scare kittens (because my app uses those), but it still works like a charm in the sample. The buttons within the collection node cells all output the `debugPrint` statement on being tapped. 

Today I integrated the sample code into my original application. Barring some changes in class names, the two are now identical. And sadly, the same code doesn't not trigger the debug output upon button touches in the app. 😢 While the active code paths are identical in the sample and the app, what does differ dramatically is the project settings and dependencies. The sample only has three dependencies (AsyncDisplayKit and its two dependencies) while the app has ten or so (Fabric, Digits, AWS iOS SDK, Alamofire, AlamofireImage, and a handful more). 

So, the $64,000 question: Anybody have **any** ideas better than, "Create a new project and spend a day or two migrating your rather large app even though that may not even fix your issue..."? 

Thanks in advance.
# pleasesaveme

p.s. thanks for ASDK; this issue notwithstanding, this framework is an incredible resource for the community (and filled a huge gap that UIKit has failed to fill for years despite their best efforts). 👏 🙌 ❤️ 
—

Version: 1.9.91 [(related commit)](https://github.com/facebook/AsyncDisplayKit/tree/a705ed09cbf18569b713f7ec036bb12a886c71ff)
Swift version: 3.0
iOS Target version: 9.0
Sample (this is the working sample; not the app): [download](https://www.dropbox.com/s/ljw9uvdrkyxvhcz/ASButtonNodeBugSample.zip?dl=0)

**Comment by appleguy** [comment]
@rainypixels I would love to help you out with this!  If you could ping me on Slack, we could work through some debugging steps.

Naturally it's important to dump a "po [[UIWindow keyWindow] recursiveDescription]", so if you could send this over (or post it here if it's not sensitive), that would be the best start.

ASDK 2 is now very stable and is better in 1.9.91 in really all respects; we're shipping Pinterest with it, Buffer is using it at scale, and others as well.  So as you mentioned in the core of your message (but not at the end), you should jump on master or one of our branches (like releases/p6.13) and debug from that state.

The most likely suspect is that you have a supernode/superview that is zero-sized or sized smaller than it should be, and one of the children is outside the bounds of that parent (which must have clipsToBounds = NO, if this theory is correct).  In such a case, display will work perfectly, but event delivery won't occur.

If you need to dive in, you can set a breakpoint on hitTest: or pointInside and follow the UIKit event delivery process.  You should be able to do this on _ASDisplayView, which will make the debugging more concrete than if you set a symbolic breakpoint on UIKit itself, and it will cut out some of the extra levels that probably aren't the issue (UIWindow, UIViewController container views, etc).

**Comment by rainypixels** [comment]
I just found the issue thanks to your `hitTest` symbolic breakpoint tip!

Embarrassingly enough, I actually found a method override with this signature in a `UICollectionViewCell` extension:

```
open override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
    ...
}
```

I remember adding it a while back to help with some debugging/touch visualization, and I must have checked it in by mistake. The method was effectively swallowing touches in `UICollectionViewCell` instances. This never affected my project until I integrated ASDK a couple of weeks ago, and converted my tableview-based feeds in favor of collectionview nodes. Perfect storm!

Thanks much, @appleguy! I'm going to go ahead and close this issue.
